### PR TITLE
First framework proposal

### PR DESCRIPTION
**Pull request by MarcCote**
@mgermain @ASalvail this is a first proposal for the training framework.

This PR proposes a modular architecture for the new machine learning framework: smartpy. The primary blocs are: `Task`, `Trainer`, `Optimizer`, "`UpdateRule`" and provides interface for `Model` and `Dataset`.

Here, an optimizer provides the way to update the parameters of a model. A trainer is more like a manager that will execute tasks in addition to tell the optimizer it is time to update the parameters once.

**Usage**
Training a model is simple as:
1. Select an `Optimizer` (e.g. `SGD`)
   - Specify what model to optimize
   - Set the loss function to optimize
   - Specify the dataset to optimize on
   - Add one or several `UpdateRule`
2. Create a `Trainer`
   - Specify the optimizer to use
   - Add one or several stopping criteria (e.g. `MaxEpochStopping`)
3. Add tasks to the trainer 
   - e.g. saving the model each N epochs (TODO)
   - e.g. early_stopping (TODO)
   - e.g. printing the classification error after each update/epoch
   - e.g. printing the gradients

To see a working example check this other PR https://github.com/SMART-Lab/models/pull/1

### CODE REVIEW

File: smartpy/core/trainer.py [hunk 1]
```diff
@@ -0,0 +1,85 @@
+from collections import OrderedDict
+
+import theano
+from time import time
+from .status import Status
+
+
+class Trainer(object):
+    def __init__(self, optimizer, status=None):
```
Comment (mgermain) [inline smartpy/core/trainer.py]: Should have a StoppingCriterion in the constructor. Training forever is not a really valid state for the trainer. It is valid only for online learning but I think it's better if it is explicit like MaxEpoch(np.inf).

File: smartpy/update_rules/learning_rate.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
+import numpy as np
+
+from collections import OrderedDict
+from smartpy.misc.utils import sharedX
+
+from . import UpdateRule
+
+
+class LearningRate(UpdateRule):
```
Comment (mgermain) [inline smartpy/update_rules/learning_rate.py]: Should probably be 2 classes. ConstantLearningRate and DecreasingLearningRate.

File: smartpy/update_rules/update_rule.py [hunk 3]
```diff
@@ -0,0 +1,13 @@
+from abc import ABCMeta, abstractmethod
+from collections import OrderedDict
+
+
+class UpdateRule(object):
+    __metaclass__ = ABCMeta
+
+    def __init__(self):
+        self.params = OrderedDict()
```
Comment (mgermain) [inline smartpy/update_rules/update_rule.py]: If this is not used right now we should remove it.

File: smartpy/optimizers/sgd.py [hunk 4]
```diff
@@ -0,0 +1,51 @@
+import theano
+import theano.tensor as T
+from collections import OrderedDict
+
+import numpy as np
+from . import Optimizer
+
+
+class SGD(Optimizer):
+    def __init__(self, model, loss_fct, dataset, batch_size=1, update_rules=None, param_modifiers=None):
```
Comment (mgermain) [inline smartpy/optimizers/sgd.py]: I don't think this is necessary in the __init__ update_rules=None, param_modifiers=None

File: smartpy/optimizers/optimizer.py [hunk 5]
```diff
@@ -0,0 +1,26 @@
+import theano.tensor as T
+
+
+class Optimizer(object):
+    def __init__(self, model, loss_fct, dataset, update_rules=None, param_modifiers=None):
```
Comment (mgermain) [inline smartpy/optimizers/optimizer.py]: I don't think this is necessary in the `__init__` update_rules=None, param_modifiers=None

File: smartpy/optimizers/optimizer.py [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+        self.dataset = dataset
+
+        self.update_rules = update_rules if update_rules is not None else []
+        self.param_modifiers = param_modifiers if param_modifiers is not None else []
+
+        self.data = [dataset.inputs_shared, dataset.targets_shared]
+        self.inputs = [T.matrix('input' + str(i)) for i in range(len(self.data))]
+        self.loss = loss_fct(*self.inputs)
+
+    def append_update_rule(self, update_rule):
```
Comment (mgermain) [inline smartpy/optimizers/optimizer.py:23]: It should have an append_param_modifiers() function too or maybe the param_modifiers and update_rules should be added by the same function, I'm not sure.

File: smartpy/optimizers/optimizer.py [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+        self.update_rules.append(update_rule)
```
Comment (mgermain) [inline smartpy/optimizers/optimizer.py]: Should be private  and there should probably be some type checking ... let's look at python standards and good practices for that.

File: smartpy/optimizers/sgd.py [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+        super(SGD, self).__init__(model, loss_fct, dataset, update_rules=update_rules, param_modifiers=param_modifiers)
+        self.batch_size = batch_size
+        self.nb_updates_per_epoch = int(np.ceil(len(dataset) / self.batch_size))
+
+    def build_learning_function(self, extra_updates={}):
```
Comment (mgermain) [inline smartpy/optimizers/sgd.py]: Private function ?

**Comment by MarcCote** [comment]
@ASalvail if you want to comment on a particular line of code, you should do it in the `File changed` tab (at the top, below the name of the PR). 

It will be easier to read because lines that have been commented on will be included directly in this discussion (right now, only the comments are been displayed). Also, modified lines (i.e. lines modified by subsequent commits in this PR) will be hidden automatically.

Thanks for the reviewing.

File: smartpy/update_rules/learning_rate.py (hunk 2 above)
Comment (MarcCote) [inline smartpy/update_rules/learning_rate.py]: Ok, I split it.

File: smartpy/core/trainer.py (hunk 1 above)
Comment (MarcCote) [inline smartpy/core/trainer.py]: Ok, I made the trainer constructor requiring a `StoppingCriterion`.

File: smartpy/update_rules/update_rule.py (hunk 3 above)
Comment (MarcCote) [inline smartpy/update_rules/update_rule.py]: Removed

File: smartpy/optimizers/optimizer.py (hunk 5 above)
Comment (MarcCote) [inline smartpy/optimizers/optimizer.py]: Removed

File: smartpy/optimizers/optimizer.py (hunk 6 above)
Comment (MarcCote) [inline smartpy/optimizers/optimizer.py:23]: For now, I put two methods, one for each.

File: smartpy/optimizers/optimizer.py (hunk 7 above)
Comment (MarcCote) [inline smartpy/optimizers/optimizer.py]: Made private. Type checking should be done in another PR.

File: smartpy/optimizers/sgd.py (hunk 8 above)
Comment (MarcCote) [inline smartpy/optimizers/sgd.py]: Yep, I changed it.

**Comment by MarcCote** [comment]
@ASalvail @mgermain I made a new commit fixing some issues you raised and porting the library to Python 3.

File: smartpy/optimizers/sgd.py [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+        updates_from_update_rules = OrderedDict()
+        for update_rule in self._update_rules:
+            gradients, updates_from_update_rule = update_rule.apply(self.gradients)
+            self.gradients.update(gradients)
+            updates_from_update_rules.update(updates_from_update_rule)
+
+        # Update parameters
+        params_updates = OrderedDict()
+        for param, gparam in self.gradients.items():
+            params_updates[param] = param - self.gradients[param]
```
Comment (ASalvail) [inline smartpy/optimizers/sgd.py]: `gradient` -> `direction`

File: smartpy/optimizers/sgd.py [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+from . import Optimizer
+
+
+class SGD(Optimizer):
+    def __init__(self, model, loss_fct, dataset, batch_size=1):
+        super(SGD, self).__init__(model, loss_fct, dataset)
+        self.batch_size = batch_size
+        self.nb_updates_per_epoch = int(np.ceil(len(dataset) / self.batch_size))
+
+    def _build_learning_function(self, extra_updates={}):
```
Comment (ASalvail) [inline smartpy/optimizers/sgd.py]: `task_updates`

File: smartpy/optimizers/sgd.py [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+        updates.update(updates_from_param_modifiers)
+        updates.update(extra_updates)
+
+        no_batch = T.iscalar('no_batch')
+        givens = {input: data[no_batch * self.batch_size:(no_batch + 1) * self.batch_size] for input, data in zip(self.inputs, self.data)}
+        learn = theano.function([no_batch],
+                                updates=updates,
+                                givens=givens,
+                                name="learn")
+        return learn
```
Comment (ASalvail) [inline smartpy/optimizers/sgd.py]: Change for `self.learn`

File: smartpy/optimizers/sgd.py [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+import theano.tensor as T
+from collections import OrderedDict
+
+import numpy as np
+from . import Optimizer
+
+
+class SGD(Optimizer):
+    def __init__(self, model, loss_fct, dataset, batch_size=1):
+        super(SGD, self).__init__(model, loss_fct, dataset)
```
Comment (ASalvail) [inline smartpy/optimizers/sgd.py:7]: Add a `Loss` object that builds the graph and can be interrogated to get the inputs variables.

File: smartpy/optimizers/sgd.py [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+        self.batch_size = batch_size
```
Comment (ASalvail) [inline smartpy/optimizers/sgd.py:8]: Make a `Batch` object that will use advanced indexing to keep track internally of the batch size.

File: smartpy/optimizers/sgd.py (hunk 11 above)
Comment (ASalvail) [inline smartpy/optimizers/sgd.py]: Return `True` when an epoch is over? Also, we could ask the `Batch` object.

File: smartpy/optimizers/sgd.py (hunk 12 above)
Comment (MarcCote) [inline smartpy/optimizers/sgd.py:7]: Will work on this in another PR. I created an issue for it. See #12

File: smartpy/optimizers/sgd.py (hunk 13 above)
Comment (MarcCote) [inline smartpy/optimizers/sgd.py:8]: Will work on this in another PR. I created an issue for it. See #13

**Comment by MarcCote** [comment]
@ASalvail @mgermain I think everyone agreed we should merge this and build upon. 

I think the next step should be to discuss about the `Loss` and `Batch` classes that would help us clean the `Optimizer`.
